pkg/spanconfig/spanconfig.go

```diff
@@ -117,6 +117,32 @@ func FullTranslate(
 	return s.Translate(ctx, descpb.IDs{keys.RootNamespaceID})
 }
 
+// SQLWatcher watches for events on system.zones and system.descriptors.
+type SQLWatcher interface {
+	// WatchForSQLUpdates watches for changes to zones and descriptors starting at
+	// the given timestamp (exclusive), informing callers using the handler
+	// callback.
+	//
+	// The handler callback is invoked from time to time with a list of updates
+	// and a checkpointTS. Invocations of the handler callback provide the
+	// following semantics:
+	// 1. Calls to the handler are serial.
+	// 2. The timestamp supplied to the handler is monotonically increasing.
+	// 3. The list of DescriptorUpdates supplied to handler includes all events
+	// in the window (prevInvocationCheckpointTS, checkpointTS].
+	// 4. No further calls to the handler are made if an invocation returns an
+	// error.
+	//
+	// These guarantees mean that users of this interface are free to persist the
+	// checkpointTS and later use it to re-establish the SQLWatcher without
+	// missing any updates.
+	WatchForSQLUpdates(
```

Comment:
I think you should do more to explain the semantics of the memory errors. The caller is going to need to deal with them and restart. That's now part of the contract.